### PR TITLE
site(Calendar): add  lunar calendar demo

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [X] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- https://github.com/ant-design/ant-design/issues/44150

### 💡 Background and solution

image
image
image
image

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Add lunar calendar demo |
| 🇨🇳 Chinese | 增加农历示例  |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at d3ad106</samp>

This pull request adds a new feature to the `Calendar` component that allows customizing the labels of the year and month select boxes with additional information. It also adds a new lunar calendar example that demonstrates this feature and uses the `lunisolar` library to get the lunar and solar term information. The pull request updates the documentation and the demo files accordingly.

### <samp>🤖 Generated by Copilot at d3ad106</samp>

*  Add lunar calendar example to the `Calendar` component ([link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-6544ea388409d5c71696262455ab086cb9e3b229f544ee0580e1edd925c00e0aR1-R7), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-ca070ff551f2a8ae4ae9890edcc6a88a39335d0d2f72648cca683e51e9923336R1-R92), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-8fab8ed47a3daf0635a3eb887b1fd4fdb984ec4daba5af51505fea4a5d87051bR23), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-58e290dfe67aeb90a9d97d4e3e8a790c9aa964113593229b7f2637b66f0c8a50R24))
* Add `getYearLabel` and `getMonthLabel` props to `Calendar` component ([link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2R76-R77), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2R119-R120), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2R338-R339), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L22-R36), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L41-R55), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L78-R107), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L100-R129), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807R180-R181), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-8fab8ed47a3daf0635a3eb887b1fd4fdb984ec4daba5af51505fea4a5d87051bR58-R59), [link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-58e290dfe67aeb90a9d97d4e3e8a790c9aa964113593229b7f2637b66f0c8a50R63-R64))
* Add `lunisolar` library as a dev dependency in `package.json` file ([link](https://github.com/ant-design/ant-design/pull/44166/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R261))

### CODE REVIEW

File: components/calendar/Header.tsx [hunk 1]
```diff
@@ -149,6 +177,8 @@ export interface CalendarHeaderProps<DateType> {
   fullscreen: boolean;
   onChange: (date: DateType, source: SelectInfo['source']) => void;
   onModeChange: (mode: CalendarMode) => void;
+  getYearLabel?: (oriLabel: string, year: number) => string;
```
Comment (afc163) [inline components/calendar/Header.tsx:180]: 这是加新特性了？最好先来个 RFC

Comment (kiner-tang) [reply components/calendar/Header.tsx:180]: 就是加了可以自定义 两个下拉框 label 的方法，方便用户可以进行一些自定义的渲染和展示
image

Comment (kiner-tang) [reply components/calendar/Header.tsx:180]: > 这是加新特性了？最好先来个 RFC

那我抽空先整理个 RFC 下周讨论一下

**Review by MadCcc — APPROVED** [review]
LGTM

**Comment by zombieJ** [comment]
留档：优化 Hover 样式

截屏2023-08-16 23 27 10

File: components/calendar/Header.tsx (hunk 1 above)
Comment (crazyair) [reply components/calendar/Header.tsx]: 需要加新特性吗，用自定义 header 可以吗，随便使用麻烦点，但更灵活

**Comment by crazyair** [comment]
image
不是右对齐了

可以把`Option`改成用 options 方式吗

**Comment by kiner-tang** [comment]
> image 不是右对齐了
> 
> 可以把`Option`改成用 options 方式吗

done

**Comment by kiner-tang** [comment]
好像选择日期的事件丢了，我改改

File: components/calendar/__tests__/__snapshots__/demo.test.ts.snap [hunk 2]
```diff
@@ -4753,6 +4753,1054 @@ exports[`renders components/calendar/demo/customize-header.tsx correctly 1`] = `
 </div>
 `;
 
+exports[`renders components/calendar/demo/lunar.tsx correctly 1`] = `
+<div
+  class="acss-5gxtu4"
```
Comment (afc163) [inline components/calendar/__tests__/__snapshots__/demo.test.ts.snap:4758]: 这些 hash 是不稳定的，放到 snapshot 里不合适吧。

Comment (kiner-tang) [reply components/calendar/__tests__/__snapshots__/demo.test.ts.snap:4758]: 那这种情况我们通常怎么处理呢

Comment (kiner-tang) [reply components/calendar/__tests__/__snapshots__/demo.test.ts.snap:4758]: 这个哈希不是根据内容生成的吗？如果样式不改变的话，hash 应该不会改变的吧

File: components/calendar/demo/lunar.tsx [hunk 3]
```diff
@@ -0,0 +1,249 @@
+import type { Dayjs } from 'dayjs';
+import dayjs from 'dayjs';
```
Comment (zombieJ) [inline components/calendar/demo/lunar.tsx:2]:
```tsx
import dayjs, { type Dayjs } from 'dayjs';
```

File: components/calendar/demo/lunar.tsx [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  const onPanelChange = (value: Dayjs, mode: CalendarMode) => {
+    console.log(value.format('YYYY-MM-DD'), mode);
+    setSelectDate(value);
+  };
+
+  const onDateChange = (value: Dayjs) => {
+    setSelectDate(value);
+  };
+
+  const cellRender = (date: Dayjs, info: CellRenderInfo<Dayjs>) => {
```
Comment (zombieJ) [inline components/calendar/demo/lunar.tsx:106]:
```tsx
const cellRender: CalendarProps['fullCellRender'] = ...
```

避免在例子中使用 `antd/lib/...` 路径

Comment (zombieJ) [reply components/calendar/demo/lunar.tsx:106]: `CalendarMode`  同样

File: components/calendar/demo/lunar.tsx [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+          [styles.today]: date.isSame(dayjs(), 'date'),
+        }),
+        children: (
+          <div className={styles.text}>
+            {date.get('date')}
+            {info.type === 'date' && <div className={styles.lunar}>{solarTerm || lunar}</div>}
+          </div>
+        ),
+      });
+      // return (
```
Comment (zombieJ) [inline components/calendar/demo/lunar.tsx:124]: 注释删除

File: components/calendar/demo/lunar.tsx [hunk 6]
```diff
@@ -0,0 +1,249 @@
+import type { Dayjs } from 'dayjs';
+import dayjs from 'dayjs';
+import React from 'react';
+import type { CellRenderInfo } from 'rc-picker/lib/interface';
```
Comment (afc163) [inline components/calendar/demo/lunar.tsx:4]: 不要依赖 rc-picker，如果需要则应该从 antd 中导出来。

File: components/calendar/demo/lunar.tsx [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+          const month = value.month();
+          const options = [];
+          for (let i = year - 10; i < year + 10; i += 1) {
+            options.push({
+              label: getYearLabel(i),
+              value: i,
+            });
+          }
+          return (
+            <Row justify="end" gutter={8} style={{ padding: '10px' }}>
```
Comment (afc163) [inline components/calendar/demo/lunar.tsx:205]:
```suggestion
            <Row justify="end" gutter={8} style={{ padding: 8 }}>
```

File: components/calendar/demo/lunar.tsx [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    monthCellCurrent: css`
+          color: ${token.colorTextLightSolid};
+          background: ${token.colorPrimary};
+          &:hover {
+            background: ${token.colorPrimary};
+            opacity: .8;
+          }
+        `,
+  };
+});
```
[the remark this answers is not in the document]
Comment (kiner-tang) [reply components/calendar/demo/lunar.tsx:88]: 主要是包含各种状态，如今天，当前选中、hover、日期面板和月份面板等一些样式需要覆盖，但每个地方要覆盖的其实不多

**Review by zombieJ — APPROVED** [review]

**Pull request merged**